app/qml/editor/inputrelation.qml

```diff
@@ -0,0 +1,268 @@
+          }
+          PropertyChanges {
+            target: noOfFeaturesText
+            visible: false
+          }
+        },
+        State {
+          name: "page"
+          // TODO: page state
+        }
+      ]
+
+      transitions: [
+        Transition {
+          from: "initial"
+          to: "expanded"
+          animations: NumberAnimation { property: "height"; duration: 100 }
+        }
+      ]
+
+      onStateChanged: delegateModel.update()
+
+      visible: mode === "text"
+      height: customStyle.fields.height
+      width: parent.width
+      state: "initial"
+
+      border.color: customStyle.fields.normalColor
+      border.width: 1 * QgsQuick.Utils.dp
+      color: customStyle.fields.backgroundColor
+      radius: customStyle.fields.cornerRadius
+
+      Text {
+        id: noOfFeaturesText
+
+        anchors.fill: parent
+        font.pointSize: customStyle.fields.fontPointSize
+        color: customStyle.fields.fontColor
+        horizontalAlignment: Qt.AlignHCenter
+        verticalAlignment: Qt.AlignVCenter
+
+        text: qsTr( "%n linked feature(s)", "Shows how many features are linked via relations", root.linkedFeaturesCount )
+      }
+
+      MouseArea {
+        anchors.fill: parent
+        onClicked: {
+          if ( textModeContainer.state === "initial" )
+            textModeContainer.state = "expanded"
+          else if ( textModeContainer.state === "expanded" && root.canOpenFeaturesPage )
+            textModeContainer.state = "page"
+        }
+      }
+
+      Flow {
+        id: flowItemView
+
+        anchors.fill: parent
+        anchors.margins: customStyle.fields.sideMargin
+
+        spacing: customStyle.group.spacing
+
+        Repeater { model: delegateModel; }
+      }
+    }
+
+    // Photo Mode Widget
+    Item {
+      id: photoModeContainer
+
+      visible: mode === "photo"
+      anchors.fill: parent
+
+      // todo: photo panel widget
+    }
+  }
+
+  Component {
+    id: textDelegate
+
+    Item {
+      id: textDelegateContainer
+
+      property real maximumWidth: flowItemView.width
+      property bool isVisible: {
+        // figure out which row am I from Y
+        if ( y === 0 ) return true
+        else if ( y < 3 * height ) return true
+        else return false
+      }
+
+      height: customStyle.relationComponent.textDelegateHeight
+      width: childrenRect.width > maximumWidth ? maximumWidth : childrenRect.width
+
+      visible: isVisible
+
+      Rectangle {
+        id: textDelegateContent
+
+        property real requestedWidth: txt.paintedWidth + 10 * QgsQuick.Utils.dp
+
+        height: parent.height
+        width: requestedWidth > parent.maximumWidth ? parent.maximumWidth : requestedWidth
+
+        color: customStyle.fields.backgroundColorDark
+        radius: customStyle.fields.cornerRadius
+        border.color: customStyle.fields.backgroundColorDarker
+        border.width: 2 * QgsQuick.Utils.dp
+
+        Text {
```

Comment:
Maybe consider elide left for long text to see beginning of the relation name. However, its just my preference.

Reply:
That is a good point, however, you can have an elide only when `Text` has width property specified, in this case it grows with `paintedWidth` and it did not work well. I will mark it down and have a look in future